### PR TITLE
Reduce complexity

### PR DESCRIPTION
**Pull request by Jean-Grimal**
I think this works : we only loop one through the withdraw queue, and for each market

I realized there was another problem in the code : in `_calcMarketSub`, the amount that can be withdrawn from a market is considered to be the available liquidity of the market. But actually it is the minimum between the liquidity of the market and the assets supplied by the vault in the market.

I fixed it quickly but without taking into both the accrued interest (since the last interaction with the vault) and the virtual shares/assets

A better fix would be to import the Morpho libraries : min(availableLiquidity, ) but it would require using the MorphoBalanceLib.

What do you think about this ?

### CODE REVIEW

**Comment by bugduino** [comment]
Thanks for you PR @Jean-Grimal.
Regarding the availableLiquidity fix I agree with you and for the accrued interest it should not be a real issue in the end, while for virtual shares/assets I'm not sure what you exactly mean here. In general I would prefer to avoid having to import MorphoBalanceLib if possible.

Regarding the loop I'm still trying to understand it better, will get back later on this

File: contracts/strategies/morpho/MetaMorphoStrategy.sol [hunk 1]
```diff
@@ -367,40 +368,24 @@ contract MetaMorphoStrategy is ERC4626Strategy {
...
+    IMorpho.Position memory _position = MORPHO_BLUE.position(_targetMarketId, address(_mmVault));
 
-    return _sub;
+    uint256 _vaultAssets = _position.supplyShares * _market.totalSupplyAssets / _market.totalSupplyShares;
+    uint256 _availableLiquidity = _market.totalSupplyAssets - _market.totalBorrowAssets;
+
+    uint256 _withdrawable = _vaultAssets > _availableLiquidity ? _vaultAssets : _availableLiquidity;
+
+    if (_sub > _withdrawable) {
+      (toSub, remaining) = (_withdrawable, _sub - remaining);
```
Comment (bugduino) [inline contracts/strategies/morpho/MetaMorphoStrategy.sol:386]: Shouldn't this be `(toSub, remaining) = (_withdrawable, _sub - _withdrawable);` ? `remaining` should be still `0` at this point

File: contracts/strategies/morpho/MetaMorphoStrategy.sol [hunk 2]
```diff
@@ -283,15 +283,16 @@ contract MetaMorphoStrategy is ERC4626Strategy {
         _aprData.add = _calcMarketAdd(_mmVault, _marketId, _aprData.supplyLen, _aprData.add);
       }
       // calc how much of `sub` will be removed from this market
+      uint256 _toSub;
       if (_aprData.sub > 0) {
-        _aprData.sub = _calcMarketSub(_mmVault, _marketId, _aprData.withdrawLen, _aprData.sub);
+        (_toSub, _aprData.sub) -= _calcMarketSub(_mmVault, _marketId, _aprData.sub);
```
Comment (bugduino) [inline contracts/strategies/morpho/MetaMorphoStrategy.sol:288]: Why the `-=`? Shouldn't this be simply `=` ?

Comment (Jean-Grimal) [reply contracts/strategies/morpho/MetaMorphoStrategy.sol:288]: Yes it should only be `=`, my bad

File: contracts/strategies/morpho/MetaMorphoStrategy.sol (hunk 1 above)
Comment (Jean-Grimal) [reply contracts/strategies/morpho/MetaMorphoStrategy.sol:386]: Yes you are right, this is a mistake

**Comment by bugduino** [comment]
I've tried to test this a bit but I found out that it cannot work in this way.
`_getRewardApr` is looping through all markets, but this is done just to find the correct market associated with the target reward, this means that all other markets are skipped. So when we call `_calcMarketSub` this would be called only for the market that has the reward associated and so it will be called only once. For this reason we still need to perform a loop that evaluates all markets for available liquidity.

Additionally I'm not sure that the amount that can be withdrawn from a market should consider the vault liquidity for that market as in theory I can withdraw all the liquidity from a single market if there is enough liquidity. Eg if we have a vault with supplyQueue = [market0, market1] and withdrawQueue = [market1, market0] if I deposit 100 underlyings and the cap is not reached the deposit will go all on market0. On withdraw, if there is at least 100 available on market1 the withdraw will happen all on market1 if I'm not missing something.

What do you think?

**Comment by MerlinEgalite** [comment]
I indeed think that it would be possible to remove one loop by doing the computation the following way:
1. Update an internal state storing the shares of the vault on each enable market after add/sub
2. Loop over the rewards
3. Internally, loop over the withdraw queue 
4. In this loop update the apr 

Maybe the cost of 1 is too expensive though.

@Jean-Grimal solution should work in theory. What is the failing test?

**Comment by bugduino** [comment]
But for 1. atm we are not actually checking all markets, but just the target market for the reward. Eg the 'idle' market has no rewards generally and this market is never touched here https://github.com/Idle-Labs/idle-tranches/pull/86/files#diff-b2b538ff3955f5fdfdd12941175e16909cf6344c74ac7170fcc268062ef4920bL271-L276 

Atm we loop for 2. then we loop internally for 3. (but here we actually read data of a single market, so it's sort of O(1)), and then we loop (for real this time) for withdrawQueue/supplyQueue to updated add/sub

**Comment by Jean-Grimal** [comment]
I went through the code. Actually there are a lot of reasons that the test doesn’t pass : 
    - Accrued Interests are not taken into account (they can be not negligible if the block number of the fork is outdated)
    - Virtual shares and amounts are not taken into account within the shares/Amount conversions
    - Withdrawals impact on the rates is not taken into account. Indeed, when assets are withdrawn from (or supplied into) a vault, the utilisation changes, and so does the borrowing rate. According to the liquidity of the rate, it can greatly affect the APY. I have the formulas of the impact of utilisation variation on rates If you need them.

But I also looked at the previous version of the code, and can’t understand how the test passes :
`_aprData.sub` is supposed to be the total amount to withdraw from the vault.
`_calcMarketSub` is supposed to compute the amount to withdraw from one market, so the minimum between the remaining amount of _aprData.sub (because some assets might have already been withdrawn from the previous markets of the withdraw queue) and the maximum withdrawable amount of the market.
But `_aprData.sub` is assigned to the return value of `_calcMarketSub`, see below :
`_aprData.sub = _calcMarketSub(_mmVault, _marketId, _aprData.withdrawLen, _aprData.sub);`
Or the amount to withdraw from one market (i.e. `_aprData.sub`) should be removed from the total amount to withdraw from the vault (`_aprData.sub`).
So we should have : 
`_aprData.sub -= _calcMarketSub(_mmVault, _marketId, _aprData.withdrawLen, _aprData.sub);`
Therefore the code is incorrect (but the new version without the loop isn’t : the return value of `_calcMarketSub` is stored in `toSub` and removed from `_aprData.sub`, as it should be).

But also, as I said before, the previous version didn’t take into account that the maximum withdrawable amount of a market is not only the available liquidity, but the minimum between the available liquidity and the assets supplied in the market by the vault.

Moreover, all the previous errors I mentioned above (missing of accrued interests, virtual shares etc) are also there in the previous version.

My only is that all these errors cancel each other out, and that’s why the previous tests used to pass.

**Comment by bugduino** [comment]
> I went through the code. Actually there are a lot of reasons that the test doesn’t pass : - Accrued Interests are not taken into account (they can be not negligible if the block number of the fork is outdated) - Virtual shares and amounts are not taken into account within the shares/Amount conversions

Interest should accrue on each interaction in theory no? but even if is not accrued for some days the apr shouldn't change much considering that in the test we have aprs in the order of 1-2% so even with 20M TVL the additional assets not considered are in the order of 20M * 1-2% / 365 ie something like 1000$ worth of underlyings per day at most.

What do you mean with virtual share and amounts are not considered? Would this change the assets of the market substantially? because otherwise it's not really relevant for our use case (similar to the interest accrued which is not considered as it won't substantially change the final rewards apr)

> 
> But I also looked at the previous version of the code, and can’t understand how the test passes : `_aprData.sub` is supposed to be the total amount to withdraw from the vault. `_calcMarketSub` is supposed to compute the amount to withdraw from one market, so the minimum between the remaining amount of _aprData.sub (because some assets might have already been withdrawn from the previous markets of the withdraw queue) and the maximum withdrawable amount of the market. But `_aprData.sub` is assigned to the return value of `_calcMarketSub`, see below : `_aprData.sub = _calcMarketSub(_mmVault, _marketId, _aprData.withdrawLen, _aprData.sub);` Or the amount to withdraw from one market (i.e. `_aprData.sub`) should be removed from the total amount to withdraw from the vault (`_aprData.sub`). So we should have : `_aprData.sub -= _calcMarketSub(_mmVault, _marketId, _aprData.withdrawLen, _aprData.sub);` Therefore the code is incorrect (but the new version without the loop isn’t : the return value of `_calcMarketSub` is stored in `toSub` and removed from `_aprData.sub`, as it should be).

That loop is run only once for the target market of the reward, so `_aprData.sub` is not reused. See https://github.com/Idle-Labs/idle-tranches/pull/86#issuecomment-1931632489 . Saying `so the minimum between the remaining amount of _aprData.sub` is not correct because the `_aprData.sub` is never updated except for the target market, because the loop in `_getRewardData` is used only to find the target market, all other market are skipped. The `_getRewardData` is doing the following:
- It considers a specific reward which is associated to a specific market and I find that market https://github.com/Idle-Labs/idle-tranches/blob/b190d3225ae6cd290939f4bbe9467edc0cea577b/contracts/strategies/morpho/MetaMorphoStrategy.sol#L271-L276
- I want to remove `_aprData.sub` from the vault and see how this impact the specific market (all other markets not considered)
- So I calculate the impact of removing `_aprData.sub`, *only* on the specific market, with `_calcMarketSub`
- This will cause the `_aprData.sub` to be updated to the value removed from this market (the updated `_aprData.sub` is not reused on other markets because the `_getRewardData` will skip all other markets)
- I'm using the updated `_aprData.sub` to perform other calculation for this and only this specific market

> 
> But also, as I said before, the previous version didn’t take into account that the maximum withdrawable amount of a market is not only the available liquidity, but the minimum between the available liquidity and the assets supplied in the market by the vault.

I added this comment before https://github.com/Idle-Labs/idle-tranches/pull/86#issuecomment-1930499186 (second paragraph) why should we cosider also the assets supplied? I should be able to withdraw eg all from 'idle' market if there is enough liquidity even if on deposit I deposited all on the 'non-idle' market

> 
> Moreover, the previous errors I mentioned above (missing of accrued interests, virtual shares) are also there in the previous version.
> 
> My only guess is that all these errors cancel each other out, and that’s why the previous tests used to pass.

I would be very odd to have these errors cancel each other out in theory, the point is that the simulated redeems and the actual redeems are equal in the prev version (with 0.001% tolerance which is perfectly fine for us) for all fuzz runs so would be very strange that all errors are disappearing on all iterations of the test 

---

It seems (but I may be wrong ofc) that the major misunderstaing is related to the fact that `_getRewardData` is not really looping through all markets but it does things only on a single market out of all the withdrawQueue because if `_marketId != _targetMarketId` we do nothing

Wdyt?

**Comment by bugduino** [comment]
> > But also, as I said before, the previous version didn’t take into account that the maximum withdrawable amount of a market is not only the available liquidity, but the minimum between the available liquidity and the assets supplied in the market by the vault.
> 
> ~I added this comment before [#86 (comment)](https://github.com/Idle-Labs/idle-tranches/pull/86#issuecomment-1930499186) (second paragraph) why should we cosider also the assets supplied? I should be able to withdraw eg all from 'idle' market if there is enough liquidity even if on deposit I deposited all on the 'non-idle' market~ UPDATE: comment below
> 

 you are right here, I was considering from a depositor pov rather than the vault pov. Indeed there should be enough deposited assets into a market to be able to redeem them. Anyway the test is still passing with the update (tested locally as I'm waiting to see if this PR should be reverted or not)

**Pull request merged**